File: 02-resilience-and-caching/backend/python-server/src/services/cache_reporte.py
import json

from db import reporte_pesado
from redis_client import get_redis

CACHE_KEY = "cache:reporte_pesado"
CACHE_TTL = 10  # segundos antes de expirar


# Caché con TTL: HIT -> devuelve guardado; MISS -> calcula, guarda y devuelve.
def get_or_calculated():
    r = get_redis()
    cached = r.get(CACHE_KEY)
    if cached is not None:
        return {"origen": "cache", "datos": json.loads(cached)}   # HIT

    datos = reporte_pesado()                                       # MISS
    r.set(CACHE_KEY, json.dumps(datos), ex=CACHE_TTL)
    return {"origen": "calculo", "datos": datos}

File: 02-resilience-and-caching/backend/node-server/src/services/cacheReporte.js
import redis from '../redis.js';
import { reportePesado } from '../db.js';

// Caché con TTL: HIT -> devuelve guardado; MISS -> calcula, guarda y devuelve.
async function getOrCalculated() {
  const cached = await redis.get('cache:reporte_pesado');
  if (cached != null) {
    return { origen: 'cache', datos: JSON.parse(cached) };   // HIT
  }

  const datos = reportePesado();                             // MISS
  await redis.set('cache:reporte_pesado', JSON.stringify(datos), { EX: 10 });
  return { origen: 'calculo', datos };
}

export default { getOrCalculated };